generator/src/FluentStageFactoryGenerator.php

```diff
@@ -0,0 +1,163 @@
+use MongoDB\CodeGenerator\Definition\GeneratorDefinition;
+use MongoDB\Model\BSONArray;
+use Nette\PhpGenerator\ClassType;
+use Nette\PhpGenerator\Method;
+use Nette\PhpGenerator\PhpNamespace;
+use Nette\PhpGenerator\TraitType;
+use RuntimeException;
+use stdClass;
+use Throwable;
+
+use function array_key_last;
+use function array_map;
+use function assert;
+use function implode;
+use function sprintf;
+
+/**
+ * Generates a fluent factory class for aggregation pipeline stages.
+ * The method definition is based on the manually edited static class
+ * that imports the stage factory trait.
+ */
+class FluentStageFactoryGenerator extends OperatorGenerator
+{
+    /**
+     * All public of this class are duplicated as instance methods of the
+     * fluent factory class.
+     */
+    private const FACTORY_CLASS = Stage::class;
+
+    public function generate(GeneratorDefinition $definition): void
+    {
+        $this->writeFile($this->createFluentFactoryClass($definition));
+    }
+
+    private function createFluentFactoryClass(GeneratorDefinition $definition): PhpNamespace
+    {
+        $namespace = new PhpNamespace($definition->namespace);
+        $class = $namespace->addClass('FluentFactory');
+
+        $namespace->addUse(StageInterface::class);
+        $namespace->addUse(FieldQueryInterface::class);
+        $namespace->addUse(Pipeline::class);
+        $namespace->addUse(Decimal128::class);
+        $namespace->addUse(Document::class);
+        $namespace->addUse(Int64::class);
+        $namespace->addUse(PackedArray::class);
+        $namespace->addUse(Serializable::class);
+        $namespace->addUse(Timestamp::class);
+        $namespace->addUse(Type::class);
+        $namespace->addUse(ArrayFieldPath::class);
+        $namespace->addUse(FieldPath::class);
+        $namespace->addUse(ResolvesToArray::class);
+        $namespace->addUse(ResolvesToObject::class);
+        $namespace->addUse(Pipeline::class);
+        $namespace->addUse(AccumulatorInterface::class);
+        $namespace->addUse(ExpressionInterface::class);
+        $namespace->addUse(Optional::class);
+        $namespace->addUse(QueryInterface::class);
+        $namespace->addUse(Sort::class);
+        $namespace->addUse(BSONArray::class);
+        $namespace->addUse(stdClass::class);
+        $namespace->addUse(self::FACTORY_CLASS);
+        $class->addProperty('pipeline')
+            ->setType('array')
+            ->setComment('@var list<StageInterface>')
+            ->setValue([]);
+        $class->addMethod('getPipeline')
+            ->setReturnType(Pipeline::class)
+            ->setBody(<<<'PHP'
+                return new Pipeline(...$this->pipeline);
+                PHP);
+
+        $staticFactory = ClassType::from(self::FACTORY_CLASS);
+        assert($staticFactory instanceof ClassType);
+        foreach ($staticFactory->getMethods() as $method) {
+            if (! $method->isPublic()) {
+                continue;
+            }
```

Comment:
Moving this check into `addMethod` would allow you to simplify the entire block:
```php
array_map(
    fn ($method) => $this->addMethod($method, $namespace, $class),
    $staticFactory->getMethods(),
);
```

Reply:
Good point. I move the `isPublic` check but keep the `foreach`.